### ISSUE TITLE
Veracode reports SQL Injection issues in Dapper for .NET 4

### ISSUE THREAD

**Issue by ganeshran** [StackExchange/dapper-dot-net#341]
Veracode is a software which runs security checks on applications and it is reporting SQL Injection issues in the SqlMapper class in Dapper. 

The lines that the issues are being reported are  718,817,857,2034

**Comment by mgravell** [comment]
Do you have a specific revision to which those numbers relate? My 817, for example, is `if(type.IsValueType())`, which is unrelated - and my 2034 is `return map(objects);`. If you can indicate which version this is from, I will happily investigate, although my hunch would be simply: Veracode is wrong. In the general case, dapper is fully parameterized. There *are* some scenarios where dapper will modify raw literals, but care is taken to ensure injection safety - for example, the sometimes-useful "treat a parameter as a literal" feature is intentionally limited to integer types to minimize the attack surface.

**Comment by ganeshran** [comment]
Hi @mgravell  thanks for your reply! I have requested additional information including the source code where these warnings are being reported.

Turns out we are using 1.13 release so the line numbers might be different. I didn't find tags which identify older releases on the repo. Is there  any other way I can check out the code at release 1.13?

**Comment by mgravell** [comment]
1.13 is bdf56330b71d90e0488c84c92a0577ed72c19d28, but: that is May 2013... I'm dubious to the benefits of over-analysing that.

718: total += cmd.ExecuteNonQuery(); - not an injection risk (the line immediately before deals with the parameters)

817: `ExecuteReader`, again, the line before deals with the parameters

857: `ExecuteReader`, the code 2 lines before deals with the parameters

2034: a wrapper method that again sets up parameters and then calls `ExecuteNonQuery`

Basically, I think the security suite you are using is simply flagging any ADO.NET execution method, expecting the user (you) to manually inspect them for appropriateness. There is nothing improper in what this code is doing; it is not exposing any SQL injection surface - it is simply: using ADO.NET.

This is a false positive by that tool.

**Comment by mgravell** [comment]
As a corollary: if the external caller does bad things, such as:

    var rows = conn.Query<Orders>(
        $"select * from Orders where Name= '{name.Text}'").ToList();

then it will undeniably be unsafe; dapper doesn't cure misuse.  The correct usage there would be:

    var rows = conn.Query<Orders>(
        "select * from Orders where Name= @name", new { name  = name.Text}).ToList();
